text/0000-kafka-topics-per-tenant.md

```diff
@@ -0,0 +1,100 @@
+- **Have A Single Topics For Each Module Area**: Module areas being Inventory, Acquisitions, Circulation etc. This shares the same pain as the previous point but to a lesser extent.
+
+
+## Detailed Explanation/Design
+### Consuming
+In [folio-kafka-wrapper](https://github.com/folio-org/folio-kafka-wrapper), a sample of a subscription pattern is defined below
+
+>folio\.Default\.\w{1,}\.DI_COMPLETED
+
+where `folio` is the cluster name, `Default` is the namespace then a wildcard to capture tenant ids followed by the event type(`DI_COMPLETED`) that the topic is supposed to contain.
+Subscriptions are created using this interface:
+```
+public static SubscriptionDefinition createSubscriptionDefinition(String env, String nameSpace, String eventType)
+```
+Modifications will be made to mark the method above as deprecated and another method created to remove the wildcard which will allow the subscription pattern to look like below
+>folio\.Default\.DI_COMPLETED
+
+Safeguards will be added to ensure that a message has a tenant id as one of its headers. It is imperative that every message has a owner.
+
+### Producing
+In folio-kafka-wrapper, a topic name is generated by this interface
+```
+public static String formatTopicName(String env, String nameSpace, String tenant, String eventType)
+```
+This will generate a sample of a subscription pattern like so:
+>folio\.Default\.tenant00001\.DI_COMPLETED
+
+Modifications will be made to mark this method as deprecated and another method is created to remove the tenant id reference. This will generate a sample of a subscription pattern like so:
+>folio\.Default\.DI_COMPLETED
+
+Deprecated methods will allow an easier transition as processes roll off tenant specific topics to single topics.
+Safeguards will be added to ensure that a message has a tenant id as one of its headers. It is imperative that every message has a owner.
+## Migration
+- Changes will need to be made in folio-kafka-wrapper to stop creating topics for each tenant.
+- FOLIO modules will need to reference the latest version of folio-kafka-wrapper.
+- During a flower release, there should be no usage occurring in a FOLIO cluster. Existing topics will be deleted. Care should be taken to ensure all messages in the topic is consumed before topic deletion.
+- FOLIO modules are instantiated, new topics are created.
+- Migration can happen on a topic basis i.e a set of topics are migrated at a time instead of all topics within a module. What most important is identifying the consumers and producers of a topic to make necessary updates to allow proper topic identification, production & consumption.
+
+
+## Risks and Drawbacks
+- FOLIO's multi-versioning scheme involving tenants using different versions of modules will cause functional issues. Multiple versions of a module will be consuming from the same topic and there are no guarantees about which module version will consume a message sourced from an incompatible version.
```

Comment:
I do not have a design to address this concern. Addressing this concern will add a lot of complexity for its minimal value. I have also heard the multi-versioning is not often used. I described some of this complexity in "Related Concerns" section in the doc.

Reply:
> I do not have a design to address this concern. Addressing this concern will add a lot of complexity for its minimal value. 

Why do you consider it to have minimal value?

To me, this design change makes this concern more likely to occur, and thus increases the value of trying to address it. What am I missing?

> I have also heard the multi-versioning is not often used. I described some of this complexity in "Related Concerns" section in the doc.

What do you mean by multi-versioning?

As I understand it, hosting providers tend to have tenants running on different releases of FOLIO and when they upgrade, they have to have multiple versions of a module running together. Have I misunderstood?

Reply:
>What do you mean by multi-versioning?

Multi-versioning refers to the multi-versioning scheme described in the section highlighted by this GitHub conversation. This is where multiple versions of a folio module exists to serve different tenants in a cluster. I don’t know if there is a official name for this feature so please let me know :)

>Why do you consider it to have minimal value?

I consider it minimal value because it is mostly used in the upgrade scenario you described which is not universal. There is a stronger case if two versions of a module are online during the most of the  life of a FOLIO cluster. This is definitely up for debate. I illustrated complexities that will need to be solved in the related section of this document as a starting point.

Multi-versioning is not addressed at the database level by any FOLIO architecture component that I am aware of. Two versions of a module typically use the same database schema. So I am attempting to maintain the same standard.

In addition, the current structure does not completely support multi versioning as well. Certain module versions are not exempt from consuming messages meant for other versions. Currently, messages are not targeted to specific versions. I am referring to compatibility of a message to be processed without error by a new version of a module. If my knowledge of FOLIO’s kafka interaction is correct, version 1 and 2 of a module will consume copies of the same message and produce two outputs that could have undesirable downstream effects. Again, the current structure organizes the messages by tenant but does not isolate. Mod-pub-sub would probably handle this better, but new development has ceased due to its approach to the publisher/subscriber model.

Reply:
> Multi-versioning refers to the multi-versioning scheme described in the section highlighted by this GitHub conversation. 

I may have missed that. Apologies.

> I don’t know if there is a official name for this feature so please let me know :)

I'm not aware of an official name.

Reply:
> I consider it minimal value because it is mostly used in the upgrade scenario you described which is not universal. There is a stronger case if two versions of a module are online during the most of the life of a FOLIO cluster. This is definitely up for debate. I illustrated complexities that will need to be solved in the related section of this document as a starting point.

I agree, I think it's worth parking this topic as one the TC should discuss and decide if we should dig into further.

> Multi-versioning is not addressed at the database level by any FOLIO architecture component that I am aware of. Two versions of a module typically use the same database schema.

I believe that's only the case whilst the Tenant API operation is ongoing. And not a general concern. 

A tenant can only have a single version of a module enabled, thus the schema is only ever used by one version or another (unless there is a bug in Okapi).

> So I am attempting to maintain the same standard.

I think this is my concern, I don't necessarily agree with trying to maintain the existing standard, especially if we consider it deficient.

> In addition, the current structure does not completely support multi versioning as well. Certain module versions are not exempt from consuming messages meant for other versions. 

That's true. That probably depends upon how disciplined the module is. I don't know how old versions of the module unsubscribe from topics during an upgrade.

> Currently, messages are not targeted to specific versions. 

We don't version the messages at all, so this cannot be achieved. Using module or interface version would undermine the publisher / subscribe relationship.

> I am referring to compatibility of a message to be processed without error by a new version of a module. 



> If my knowledge of FOLIO’s kafka interaction is correct, version 1 and 2 of a module will consume copies of the same message and produce two outputs that could have undesirable downstream effects. 

I agree that is a concern and a limitation of our current design. Our disagreement may well come down to my own biases about the sequence in which I would want to address the various limitations in the design of this area of FOLIO.

> Again, the current structure organizes the messages by tenant but does not isolate. 

What do you mean in this context by does not isolate? I get the impression you have a different standard for isolation than I do.

Reply:
>I think this is my concern, I don't necessarily agree with trying to maintain the existing standard, especially if we consider it deficient.

I agree that this is our sticking point. 

I have added an open question regarding support for the multi-versioning scheme in Kafka going forward.